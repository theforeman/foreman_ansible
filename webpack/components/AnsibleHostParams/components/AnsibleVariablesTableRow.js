import React, { useState } from 'react';
import { lowerCase } from 'lodash';
import PropTypes from 'prop-types';

import AnsibleVariableInput from './AnsibleVariableInput';
import {
  overrideFieldName,
  updateFieldDisabled,
  fieldElement,
  constructId,
  roleNameColumn,
  initLookupValue,
  sanitizeFormValue,
  sanitizeFieldValue,
  lookupKeyValidations,
  shouldDestroy,
  shouldAddFields,
} from './AnsibleVariableHelpers';

const AnsibleVariablesTableRow = ({
  lookupKey,
  resourceError,
  role,
  firstKey,
  formObject,
}) => {
  const [lookupValue] = useState(initLookupValue(lookupKey, formObject));
  const [dirty, setDirty] = useState(false);
  const [fieldOverriden, setFieldOverriden] = useState(
    resourceError ? true : lookupValue.overriden
  );
  const [fieldOmmited, setFieldOmmited] = useState(
    resourceError ? false : lookupValue.omit
  );
  const [fieldValue, setFieldValue] = useState(
    resourceError ? resourceError.value : sanitizeFieldValue(lookupValue)
  );

  const [fieldDisabled, setFieldDisabled] = useState(
    resourceError
      ? false
      : updateFieldDisabled(lookupValue.overriden, lookupValue.omit)
  );
  const [fieldHiddenValue, setFieldHiddenValue] = useState(
    lookupKey.hiddenValue
  );

  const toggleOverride = () => {
    const isOverriden = !fieldOverriden;
    setFieldOverriden(isOverriden);

    setFieldValue(
      sanitizeFormValue(isOverriden ? fieldValue : lookupValue.value)
    );
    setFieldDisabled(updateFieldDisabled(!fieldOverriden, fieldOmmited));
  };

  const toggleOmit = () => {
    setFieldOmmited(!fieldOmmited);
    setFieldDisabled(updateFieldDisabled(fieldOverriden, !fieldOmmited));
  };

  const toggleHidden = () => setFieldHiddenValue(!fieldHiddenValue);

  const updateFieldValue = value => {
    setDirty(true);
    setFieldValue(value);
  };

  const keyErrors = lookupKeyValidations(
    lookupKey,
    fieldValue,
    fieldDisabled,
    resourceError,
    dirty
  );

  const formModelField = fieldElement(lowerCase(formObject.resourceName));

  return (
    <tr
      id={constructId(role, lookupKey)}
      className="fields overriden"
      key={lookupKey.id}
    >
      {firstKey && lookupKey.id === firstKey.id ? roleNameColumn(role) : null}
      <td className="elipsis param_name">{lookupKey.parameter}</td>
      <td className="elipsis">{lookupKey.parameterType}</td>
      <td className={!keyErrors.valid && !fieldDisabled ? 'has-error' : ''}>
        <AnsibleVariableInput
          role={role}
          lookupKey={lookupKey}
          lookupValue={lookupValue}
          updateFieldValue={updateFieldValue}
          toggleOverride={toggleOverride}
          toggleHidden={toggleHidden}
          fieldDisabled={fieldDisabled}
          fieldOverriden={fieldOverriden}
          fieldOmmited={fieldOmmited}
          fieldHiddenValue={fieldHiddenValue}
          fieldValue={fieldValue}
          formModelField={formModelField}
          keyErrors={keyErrors}
        />
        <span className="help-block">{!fieldDisabled && keyErrors.msg}</span>
      </td>
      <td className="ca">
        <input
          type="checkbox"
          onChange={toggleOmit}
          checked={fieldOmmited}
          value={setFieldOmmited ? '1' : '0'}
          hidden={!fieldOverriden ? 'hidden' : undefined}
          name={overrideFieldName(lookupKey, 'omit', formModelField)}
        />

        <input
          type="hidden"
          name={overrideFieldName(lookupKey, 'lookup_key_id', formModelField)}
          value={lookupKey.id}
          disabled={!shouldAddFields(lookupValue, fieldDisabled, fieldOmmited)}
        />
        <input
          type="hidden"
          name={overrideFieldName(lookupKey, 'id', formModelField)}
          value={lookupValue.id}
          disabled={!shouldAddFields(lookupValue, fieldDisabled, fieldOmmited)}
        />
        <input
          type="hidden"
          name={overrideFieldName(lookupKey, '_destroy', formModelField)}
          value={shouldDestroy(lookupValue, fieldDisabled, fieldOmmited)}
          disabled={!shouldAddFields(lookupValue, fieldDisabled, fieldOmmited)}
        />
      </td>
    </tr>
  );
};

AnsibleVariablesTableRow.propTypes = {
  lookupKey: PropTypes.object.isRequired,
  resourceError: PropTypes.object,
  role: PropTypes.object.isRequired,
  firstKey: PropTypes.object.isRequired,
  formObject: PropTypes.object.isRequired,
};

AnsibleVariablesTableRow.defaultProps = {
  resourceError: null,
};

export default AnsibleVariablesTableRow;
