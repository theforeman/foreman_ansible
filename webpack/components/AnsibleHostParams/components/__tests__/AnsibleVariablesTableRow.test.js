import { testComponentSnapshotsWithFixtures } from '@theforeman/test';

import AnsibleVariablesTableRow from '../AnsibleVariablesTableRow';

import {
  roles,
  rectangleLookupKey,
  squareLookupKey,
} from '../../AnsibleHostParams.fixtures';

const formObject = {
  resourceName: 'hostgroup',
  resourceId: null,
  parentId: null,
};

const fixtures = {
  'should render first row with a lookup key': {
    role: roles[0],
    formObject,
    firstKey: rectangleLookupKey,
    resourceError: null,
    lookupKey: rectangleLookupKey,
  },
  'should render second row with a lookup key': {
    role: roles[0],
    formObject,
    firstKey: rectangleLookupKey,
    resourceError: null,
    lookupKey: squareLookupKey,
  },
};

describe('AnsibleVariablesTableRow', () =>
  testComponentSnapshotsWithFixtures(AnsibleVariablesTableRow, fixtures));
