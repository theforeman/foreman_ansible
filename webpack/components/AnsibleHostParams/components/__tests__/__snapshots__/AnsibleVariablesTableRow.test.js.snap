// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnsibleVariablesTableRow should render first row with a lookup key 1`] = `
<tr
  className="fields overriden"
  id="ansible_role_3_params[66]"
  key="66"
>
  <td
    className="elipsis"
    rowSpan={2}
  >
    aardvark.cube
  </td>
  <td
    className="elipsis param_name"
  >
    rectangle
  </td>
  <td
    className="elipsis"
  >
    integer
  </td>
  <td
    className=""
  >
    <AnsibleVariableInput
      fieldDisabled={true}
      fieldHiddenValue={false}
      fieldOmmited={true}
      fieldOverriden={true}
      fieldValue={17}
      formModelField="hostgroup"
      keyErrors={
        Object {
          "icon": "info",
          "text": "",
          "valid": true,
        }
      }
      lookupKey={
        Object {
          "currentOverride": Object {
            "element": "hostgroup",
            "elementName": "first hostgroup",
            "value": null,
          },
          "defaultValue": 17,
          "description": "Rectangle param",
          "hiddenValue": false,
          "id": 66,
          "omit": null,
          "override": true,
          "overrideValues": Array [
            Object {
              "id": 31,
              "match": "hostgroup=first hostgroup",
              "omit": true,
              "value": null,
            },
          ],
          "parameter": "rectangle",
          "parameterType": "integer",
        }
      }
      lookupValue={
        Object {
          "defaultValue": 17,
          "element": "hostgroup",
          "elementName": "first hostgroup",
          "id": 31,
          "omit": true,
          "overriden": true,
          "value": null,
        }
      }
      role={
        Object {
          "ansibleVariables": Array [
            Object {
              "currentOverride": Object {
                "element": "hostgroup",
                "elementName": "first hostgroup",
                "value": null,
              },
              "defaultValue": 17,
              "description": "Rectangle param",
              "hiddenValue": false,
              "id": 66,
              "omit": null,
              "override": true,
              "overrideValues": Array [
                Object {
                  "id": 31,
                  "match": "hostgroup=first hostgroup",
                  "omit": true,
                  "value": null,
                },
              ],
              "parameter": "rectangle",
              "parameterType": "integer",
            },
            Object {
              "currentOverride": Object {
                "element": "hostgroup",
                "elementName": "first hostgroup",
                "value": false,
              },
              "defaultValue": true,
              "description": "Square param",
              "hiddenValue": false,
              "id": 65,
              "omit": null,
              "override": true,
              "overrideValues": Array [
                Object {
                  "id": 26,
                  "match": "hostgroup=first hostgroup",
                  "omit": false,
                  "value": false,
                },
              ],
              "parameter": "square",
              "parameterType": "boolean",
            },
          ],
          "id": 3,
          "name": "aardvark.cube",
        }
      }
      toggleHidden={[Function]}
      toggleOverride={[Function]}
      updateFieldValue={[Function]}
    />
    <span
      className="help-block"
    />
  </td>
  <td
    className="ca"
  >
    <input
      checked={true}
      name="hostgroup[lookup_values_attributes][66][omit]"
      onChange={[Function]}
      type="checkbox"
      value="1"
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][66][lookup_key_id]"
      type="hidden"
      value={66}
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][66][id]"
      type="hidden"
      value={31}
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][66][_destroy]"
      type="hidden"
      value={false}
    />
  </td>
</tr>
`;

exports[`AnsibleVariablesTableRow should render second row with a lookup key 1`] = `
<tr
  className="fields overriden"
  id="ansible_role_3_params[65]"
  key="65"
>
  <td
    className="elipsis param_name"
  >
    square
  </td>
  <td
    className="elipsis"
  >
    boolean
  </td>
  <td
    className=""
  >
    <AnsibleVariableInput
      fieldDisabled={false}
      fieldHiddenValue={false}
      fieldOmmited={false}
      fieldOverriden={true}
      fieldValue={true}
      formModelField="hostgroup"
      keyErrors={
        Object {
          "icon": "info",
          "text": "",
          "valid": true,
        }
      }
      lookupKey={
        Object {
          "currentOverride": Object {
            "element": "hostgroup",
            "elementName": "first hostgroup",
            "value": false,
          },
          "defaultValue": true,
          "description": "Square param",
          "hiddenValue": false,
          "id": 65,
          "omit": null,
          "override": true,
          "overrideValues": Array [
            Object {
              "id": 26,
              "match": "hostgroup=first hostgroup",
              "omit": false,
              "value": false,
            },
          ],
          "parameter": "square",
          "parameterType": "boolean",
        }
      }
      lookupValue={
        Object {
          "defaultValue": true,
          "element": "hostgroup",
          "elementName": "first hostgroup",
          "id": 26,
          "omit": false,
          "overriden": true,
          "value": false,
        }
      }
      role={
        Object {
          "ansibleVariables": Array [
            Object {
              "currentOverride": Object {
                "element": "hostgroup",
                "elementName": "first hostgroup",
                "value": null,
              },
              "defaultValue": 17,
              "description": "Rectangle param",
              "hiddenValue": false,
              "id": 66,
              "omit": null,
              "override": true,
              "overrideValues": Array [
                Object {
                  "id": 31,
                  "match": "hostgroup=first hostgroup",
                  "omit": true,
                  "value": null,
                },
              ],
              "parameter": "rectangle",
              "parameterType": "integer",
            },
            Object {
              "currentOverride": Object {
                "element": "hostgroup",
                "elementName": "first hostgroup",
                "value": false,
              },
              "defaultValue": true,
              "description": "Square param",
              "hiddenValue": false,
              "id": 65,
              "omit": null,
              "override": true,
              "overrideValues": Array [
                Object {
                  "id": 26,
                  "match": "hostgroup=first hostgroup",
                  "omit": false,
                  "value": false,
                },
              ],
              "parameter": "square",
              "parameterType": "boolean",
            },
          ],
          "id": 3,
          "name": "aardvark.cube",
        }
      }
      toggleHidden={[Function]}
      toggleOverride={[Function]}
      updateFieldValue={[Function]}
    />
    <span
      className="help-block"
    />
  </td>
  <td
    className="ca"
  >
    <input
      checked={false}
      name="hostgroup[lookup_values_attributes][65][omit]"
      onChange={[Function]}
      type="checkbox"
      value="1"
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][65][lookup_key_id]"
      type="hidden"
      value={65}
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][65][id]"
      type="hidden"
      value={26}
    />
    <input
      disabled={false}
      name="hostgroup[lookup_values_attributes][65][_destroy]"
      type="hidden"
      value={false}
    />
  </td>
</tr>
`;
